### PR TITLE
Introduce ProGuard integration for Compose/Desktop packaging

### LINKED ISSUE #1174: Document how to use ProGuard with Compose for Desktop
I build my binary with `packageDmg` / `packageMsi`  and I'm looking for a sample how I must modify my `build.gradle.kts`  to include an obfuscation step. I don't find an answer to that.

I think it would be nice if the official template would just contain such a step because it's a hard requirement for anybody who ships business software to customers.

In issue #607 this was asked, but the answer was using `packageUberJarForCurrentOS` and obfuscate that.
I find that not really satisfying, because I want the DMG / MSI as a result.

Further the Gradle task for Android comes with easy proguard support.
Why shoul the `nativeDistributions` section not contain a `proguardConfFile` param just to make it work out of the box?

### PR DESCRIPTION
**Pull request by AlexeyTsvetkov**
Resolves #1174

### CODE REVIEW

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationBuildTypes.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.gradle.api.Action
+import org.gradle.api.model.ObjectFactory
+import org.jetbrains.compose.desktop.application.internal.new
+import javax.inject.Inject
+
+abstract class JvmApplicationBuildTypes @Inject constructor(
+    objects: ObjectFactory
+) {
+    internal val default: JvmApplicationBuildType = objects.new("")
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationBuildTypes.kt:16]: Why empty String?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/ProguardSettings.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.jetbrains.compose.desktop.application.internal.notNullProperty
+import org.jetbrains.compose.desktop.application.internal.nullableProperty
+import javax.inject.Inject
+
+abstract class ProguardSettings @Inject constructor(
+    objects: ObjectFactory,
+) {
+    val version: Property<String> = objects.notNullProperty("7.2.2")
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/ProguardSettings.kt:18]: What means 7.2.2?
It's default version of proguard?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/ExternalToolRunner.kt [hunk 3]
```diff
@@ -46,6 +51,12 @@ internal class ExternalToolRunner(
                     // check exit value later
                     spec.isIgnoreExitValue = true
 
+                    @Suppress("NAME_SHADOWING")
+                    val logToConsole = when (logToConsole) {
+                        LogToConsole.Always -> true
+                        LogToConsole.Never -> false
+                        LogToConsole.OnlyWhenVerbose -> verbose.get()
+                    }
                     if (logToConsole) {
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/ExternalToolRunner.kt:60]:
```suggestion
                    val isLogToConsole = when (logToConsole) {
                        LogToConsole.Always -> true
                        LogToConsole.Never -> false
                        LogToConsole.OnlyWhenVerbose -> verbose.get()
                    }
                    if (isLogToConsole) {
```

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    internal val mainJarInDestinationDir: Provider<RegularFile> = mainJar.flatMap {
+        destinationDir.file(it.asFile.name)
+    }
+
+    @get:InputFiles
+    val configurationFiles: ConfigurableFileCollection = objects.fileCollection()
+
+    // todo: DSL for excluding default rules
+    // also consider pulling coroutines rules from coroutines artifact
+    // https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/jvm/resources/META-INF/proguard/coroutines.pro
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt:45]: This //todo comment needs for the future?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        val args = arrayListOf<String>().apply {
+            val maxHeapSize = maxHeapSize.orNull
+            if (maxHeapSize != null) {
+                add("-Xmx:$maxHeapSize")
+            }
+            cliArg("-cp", proguardFiles.map { it.normalizedPath() }.joinToString(File.pathSeparator))
+            add("proguard.ProGuard")
+            // todo: consider separate flag
+            cliArg("-verbose", verbose)
+            cliArg("-include", rootConfigurationFile)
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt:135]: this //todo comment needs for the future?

File: gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/gradle/DesktopApplicationTest.kt [hunk 6]
```diff
@@ -61,6 +61,21 @@ class DesktopApplicationTest : GradlePluginTestBase() {
...
+    @Test
+    fun proguard(): Unit = with(testProject(TestProjects.proguard)) {
+        gradle(":runReleaseDistributable").build().checks { check ->
+            check.taskOutcome(":proguardReleaseJars", TaskOutcome.SUCCESS)
+
+            assertEqualTextFiles(file("main-methods.actual.txt"), file("main-methods.expected.txt"))
+
+            val actualMainImage = file("main-image.actual.png")
+            val expectedMainImage = file("main-image.expected.png")
+            assert(actualMainImage.readBytes().contentEquals(expectedMainImage.readBytes())) {
```
Comment (dima-avdeev-jb) [inline gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/gradle/DesktopApplicationTest.kt:73]: How this test works?
Does "main-image.actual.png" generated in this test?

Comment (dima-avdeev-jb) [reply gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/gradle/DesktopApplicationTest.kt:73]: I already found what how this png generated

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/ProguardSettings.kt (hunk 2 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/ProguardSettings.kt:18]: Yes, I will add a constant instead

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationBuildTypes.kt (hunk 1 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationBuildTypes.kt:16]: The default build type.
It is similar concept to android debug/release build profiles.
A build profile's name is added to task's name (the pattern is `<taskAction><buildTypeName><taskSubject>`).
I don't want to break compatibility with the existing plugin by effectively renaming e.g. `packageDmg` to `packageDebugDmg`. Treating debug/dev profile as default profile without name makes sense to me.
What do you think, @dima-avdeev-jb ?

Comment (dima-avdeev-jb) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationBuildTypes.kt:16]: Yeah, good decision to save backward compatibility!

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationData.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.ProviderFactory
+import org.jetbrains.compose.desktop.application.dsl.JvmApplicationDistributions
+import org.jetbrains.compose.desktop.application.dsl.JvmApplicationBuildTypes
+import javax.inject.Inject
+
+internal open class JvmApplicationData @Inject constructor(
+    objects: ObjectFactory,
+    private val providers: ProviderFactory
```
[the remark this answers is not in the document]
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationData.kt:19]: `objects` is not a property, it is just a constructor parameter, so it does not have a visibility

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationContext.kt [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            project.mppExt.targets.all { target ->
+                if (target.platformType == KotlinPlatformType.jvm) {
+                    if (!isJvmTargetConfigured) {
+                        appInternal.from(target)
+                        isJvmTargetConfigured = true
+                    } else {
+                        project.logger.error("w: Default configuration for Compose Desktop Application is disabled: " +
+                                "multiple Kotlin JVM targets definitions are detected. " +
+                                "Specify, which target to use by using `compose.desktop.application.from(kotlinMppTarget)`")
+                        appInternal.disableDefaultConfiguration()
```
[the remark this answers is not in the document]
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationContext.kt:67]: Maybe, but this PR has not changed that, so I suppose it is out of scope

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationData.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationData.kt:19]: Yeah, sure :)

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationContext.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationContext.kt:67]: Ok, let's stay it as is

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt (hunk 5 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractProguardTask.kt:135]: Yes, that's stuff to consider

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationInternal.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import org.jetbrains.compose.desktop.application.dsl.JvmApplication
+import org.jetbrains.compose.desktop.application.dsl.JvmApplicationDistributions
+import org.jetbrains.compose.desktop.application.dsl.JvmApplicationBuildTypes
+import org.jetbrains.kotlin.gradle.plugin.KotlinTarget
+import org.jetbrains.kotlin.gradle.targets.jvm.KotlinJvmTarget
+import javax.inject.Inject
+
+internal open class JvmApplicationInternal @Inject constructor(
+    @Suppress("unused")
+    val name: String,
```
[the remark this answers is not in the document]
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/JvmApplicationInternal.kt:22]: It is actually used, I will remove the supress annotation

**Comment by dima-avdeev-jb** [comment]
All good, Thanks!

**Comment by StefanOltmann** [comment]
Really awesome that you did this! 🥇

**Comment by natanfudge** [comment]
Why proguard over R8? R8 has Kotlin-specific optimizations that should work for the desktop JVM as well.

**Pull request merged**